### PR TITLE
Simplify and fix sprite atlas coordinate calculations

### LINKED ISSUE #6863: Investigate pixelated symbols in labels
**Platform:** Ubuntu 16.04 with NVIDIA closed-source GPU driver version 367.27
**Mapbox SDK version:** master

### Steps to trigger behavior

When rendering tiles from `mapbox/streets-v8` in `MapMode::Still`, some label icons appear pixelated.

- London `mapbox/streets-v8/tiles/17/65502/43585@2x`
  - `center: [ -0.092010498046875, 51.50617819737356 ]`
- New York City `mapbox/streets-v8/tiles/17/38592/49273@2x`
  - `center: [ -74.00253295898438, 40.72748642299778 ]`
- Changing the decimal precision of the lat/lng centerpoint can affect the artifacts, sometimes fixing them if precision is trimmed enough. However, trimming centerpoint precision does not fix all instances of this artifact, and in some cases makes it worse. **Is there a potential float precision issue here and if so how should it be handled?**

**master**

43585 2x 3

**trimmed to 8 decimal places**

43585 2x 2

- [Removing `GL_LINEAR` mode rendering](https://github.com/mapbox/mapbox-gl-native/pull/6790) changes the style of artifact, fixing some but introduces others (fuzziness visible in `RBBTS` icon).

1f7d29b4-9f71-11e6-8fd7-717bf4c70bec

### Expected behavior

Label icons are rendered without pixelation or fuzziness artifacts.

bec71b9c-9f71-11e6-868b-69ffb17f86e5

### Actual behavior

Label icons are pixelated (`Villa Mosconi`, `Emmet's`, `The Dutch`).

49273 2x

/cc @jfirebaugh @brunoabinader @kkaefer @samanpwbb

### PR DESCRIPTION
**Pull request by jfirebaugh**
* Always return image metrics exclusive of padding
* Work with integer coordinates whenever possible
* Eliminate redundant SpriteAtlasElement members
* Fix asymmetric re-padding in getIconQuad when pixelRatio != 1 (fixes #6863)
* Add explanatory comments

### CODE REVIEW

**Comment by ansis** [comment]
Does this have any affect on icons with a non-integer size (https://github.com/mapbox/mapbox-gl-native/pull/3561)? I'm not seeing anything that would, but I thought I'd ask since I've missed things like this before.

This looks good, but I'm a bit rusty on all the SpriteAtlas edge cases. How deep of a review do you want on this?

**Comment by jfirebaugh** [comment]
I'll add an integration test for non-integer pixel ratio and see.

The most helpful thing would be if you could look at the [quads.cpp](https://github.com/mapbox/mapbox-gl-native/pull/9038/files#diff-a64a1fefa7d3933bc17046aee321dc88) changes and confirm the comment there. The metrics that the SpriteAtlas provides are now exclusive of padding, and it seems like the quad calculations should also be exclusive of padding, creating a quad with the same intrinsic size as the original image, whose vertex attributes contain the texture coordinates of the image exclusive of padding. That would be much simpler, but cause changes to rendering results. Is there a subtlety about the the current behavior I'm missing?

**Comment by ansis** [comment]
> The metrics that the SpriteAtlas provides are now exclusive of padding, and it seems like the quad calculations should also be exclusive of padding, creating a quad with the same intrinsic size as the original image, whose vertex attributes contain the texture coordinates of the image exclusive of padding. That would be much simpler, but cause changes to rendering results. Is there a subtlety about the the current behavior I'm missing?

If you have a 10px icon that isn't perfectly aligned to the pixel grid it will cover 11 actual pixels. The quad needs to be padded to account for this, otherwise they'll look slightly clipped on one edge in some cases.

**Comment by ansis** [comment]
So, the quad needs to be padded and the texture coordinates added to the buffers need to be padded. It's fine if SpriteAtlas returns exact coordinates as long as the padding happens elsewhere.

Since this has been very unclear and the padding is split across a bunch of different places I wouldn't be surprised if there was an off by one in terms of centering somewhere.

File: src/mbgl/text/quads.cpp [hunk 1]
```diff
@@ -21,14 +21,19 @@ SymbolQuad getIconQuad(const Anchor& anchor,
...
     const float border = 1.0;
-    auto left = shapedIcon.left() - border;
-    auto right = left + image.pos.w / image.relativePixelRatio;
-    auto top = shapedIcon.top() - border;
-    auto bottom = top + image.pos.h / image.relativePixelRatio;
+
+    float left = shapedIcon.left() - border;
+    float right = left + image.size[0] + 2 * border * pixelRatio / image.pixelRatio;
+    float top = shapedIcon.top() - border;
+    float bottom = top + image.size[1] + 2 * border * pixelRatio / image.pixelRatio;
```
Comment (ansis) [inline src/mbgl/text/quads.cpp]: I think some border-based adjustments are needed (https://github.com/mapbox/mapbox-gl-native/pull/9038#issuecomment-302761272) but multiplying border by the pixel ratio on one side and not on the other looks suspicious.

Comment (ansis) [inline src/mbgl/text/quads.cpp]: It looks like this behavior exists in -native and -js before this pull request.

Comment (ansis) [inline src/mbgl/text/quads.cpp]: if there is something wrong here, it could be the cause of https://github.com/mapbox/mapbox-gl-native/issues/6863

Comment (jfirebaugh) [inline src/mbgl/text/quads.cpp]: Here's how I would simplify this, based on the above:

```
diff --git a/src/mbgl/sprite/sprite_atlas.cpp b/src/mbgl/sprite/sprite_atlas.cpp
index c29b91797..e3df6864c 100644
--- a/src/mbgl/sprite/sprite_atlas.cpp
+++ b/src/mbgl/sprite/sprite_atlas.cpp
@@ -36,7 +36,7 @@ SpriteAtlasElement::SpriteAtlasElement(Rect<uint16_t> rect,
 SpriteAtlas::SpriteAtlas(Size size, float pixelRatio)
     : pixelSize(std::ceil(size.width * pixelRatio),
                 std::ceil(size.height * pixelRatio)),
-      padding(pixelRatio),
+      padding(1),
       bin(pixelSize.width, pixelSize.height),
       dirty(true) {
 }
diff --git a/src/mbgl/text/quads.cpp b/src/mbgl/text/quads.cpp
index 26ee237d6..2a6e2d772 100644
--- a/src/mbgl/text/quads.cpp
+++ b/src/mbgl/text/quads.cpp
@@ -22,21 +22,18 @@ SymbolQuad getIconQuad(const Anchor& anchor,
                        const float layoutTextSize,
                        const style::SymbolPlacementType placement, 
                        const Shaping& shapedText,
-                       const float pixelRatio) {
+                       const float) {
     const SpriteAtlasElement& image = shapedIcon.image();
 
     // If you have a 10px icon that isn't perfectly aligned to the pixel grid it will cover 11 actual
     // pixels. The quad needs to be padded to account for this, otherwise they'll look slightly clipped
     // on one edge in some cases.
-    //
-    // TODO: These calculations are more complex than they probably need to be, and may conceal off-by-one
-    // errors.
     const float border = 1.0;
 
-    float left = shapedIcon.left() - border;
-    float right = left + image.size[0] + 2 * border * pixelRatio / image.pixelRatio;
+    float left = shapedIcon.left() - border / image.pixelRatio;
+    float right = left + image.size[0] + 2 * border / image.pixelRatio;
     float top = shapedIcon.top() - border;
-    float bottom = top + image.size[1] + 2 * border * pixelRatio / image.pixelRatio;
+    float bottom = top + image.size[1] + 2 * border / image.pixelRatio;
     Point<float> tl;
     Point<float> tr;
     Point<float> br;
@@ -101,10 +98,10 @@ SymbolQuad getIconQuad(const Anchor& anchor,
     }
 
     Rect<uint16_t> textureRect {
-        static_cast<uint16_t>(image.tl[0] - border * pixelRatio),
-        static_cast<uint16_t>(image.tl[1] - border * pixelRatio),
-        static_cast<uint16_t>(image.br[0] - image.tl[0] + 2 * border * pixelRatio),
-        static_cast<uint16_t>(image.br[1] - image.tl[1] + 2 * border * pixelRatio)
+        static_cast<uint16_t>(image.tl[0] - border),
+        static_cast<uint16_t>(image.tl[1] - border),
+        static_cast<uint16_t>(image.br[0] - image.tl[0] + 2 * border),
+        static_cast<uint16_t>(image.br[1] - image.tl[1] + 2 * border)
     };
 
     return SymbolQuad { tl, tr, bl, br, textureRect, 0, 0, anchor.point, globalMinScale, std::numeric_limits<float>::infinity(), shapedText.writingMode };
```

Here are the rendering results affected by that change: https://fiddle.jshell.net/0n15e6nu/show/light/.

Comment (jfirebaugh) [inline src/mbgl/text/quads.cpp]: Pretty sure now that this is indeed the cause of #6863, because the change only affects non-1x renders, and #6863 has been reported only on 2x renders.

File: src/mbgl/renderer/painters/painter_symbol.cpp [hunk 2]
```diff
@@ -68,13 +68,11 @@ void Painter::renderSymbol(PaintParameters& parameters,
         auto paintPropertyValues = layer.iconPaintProperties();
 
         SpriteAtlas& atlas = *bucket.spriteAtlas;
-        const bool iconScaled = layout.get<IconSize>().constantOr(1.0) != 1.0 ||
-            frame.pixelRatio != atlas.getPixelRatio() ||
```
Comment (anandthakker) [inline src/mbgl/renderer/painters/painter_symbol.cpp]: Is this condition now being covered [here](https://github.com/mapbox/mapbox-gl-native/pull/9038/commits/67d5254f71fb643036dcf45674fb477bceed219b#diff-f27a3a157c549a13f8fe1ee8c8157ea8R274)?

Comment (jfirebaugh) [inline src/mbgl/renderer/painters/painter_symbol.cpp]: Yes, this condition was redundant and misleading. `SpriteAtlas` itself does not have any intrinsic pixel ratio; it has a collection of images of possibly heterogenous pixel ratios.
